### PR TITLE
Extra posthog attributes

### PR DESCRIPTION
**Pull request by MrZammler**
##### Summary

This PR adds a few more attributes to posthog, and makes a few changes.

Changes:

* Moves `set_late_global_environment()` before starting any threads to make sure that any events from other threads (e.g. exporting) carries those attributes.
* Provides the full agent version to analytics, instead of the short string of just the major version.
* Changes the first `META` event to be `META_START`

Adds:

* `netdata_is_in_ci`: To indicate whether this agent is in a CI environment.
* `config_oom_score`: The OOM score of the agent (assuming it can be read, and set).
* `netdata_install_type`: Tries to read the string from the `.install-type` file, in /etc/. If it doesn't exist, it will be set to an empty string.
* `config_private_registry_used`: If the user has specified a different than the default in `registry to announce`.
* `config_is_private_registry`: If the user has enabled private registry (Note: Does it also need to have a custom `registry to announce`?
* `config_https_available`: A list of three values, always on the same position, i.e. `exporting|streaming|web`. Each one will be there depending on the following:

If a user has added ssl keys in `/etc/` `ssl` folder, and the agent is able to use them, `web` will appear.
If along with that the user has enabled ssl for streaming, and it has been negotiated with the parent, `streaming` will appear.
If a user along with the keys has enabled ssl for exporting (I understand only some engines support it, if at least one has it enabled) `exporting` will appear in the string.

##### Component Name

Analytics

##### Test Plan

Compile with `NETDATA_INTERNAL_CHECKS`, use the debug flag of `0x0000000080000000` and check the debug log if attributes are correctly picked up.

##### Additional Information

We'll leave this PR as a draft at the moment, to discuss the keys with @andrewm4894 and include a change to detect ACLK Legacy/NG, as soon as the changes for building both go into master.

### CODE REVIEW

**Comment by andrewm4894** [comment]
looking at https://github.com/netdata/netdata/pull/11157 i see a line for `PREBUILT_DISTRO` i'm thinking it might be good to capture that too into something like `netdata_prebuilt_distro` property in PostHog. I think that might maybe be useful for slicing things by when we have bugs or issues that come up and help narrow down the search space a little.

File: daemon/anonymous-statistics.sh.in
Comment (ilyam8) [inline daemon/anonymous-statistics.sh.in:26]: @Ferroin 

Will the node's environment variables be propagated to the docker container?

Comment (ilyam8) [reply daemon/anonymous-statistics.sh.in:26]: I guess not and that means we have problems with `NETDATA_IS_IN_CI`

Comment (MrZammler) [reply daemon/anonymous-statistics.sh.in:26]: Hmm, hi Ilya, should we remove it then.... Or is there another way we can get the variable... ?

Comment (ilyam8) [reply daemon/anonymous-statistics.sh.in:26]: I don't see how `$CI` can be set when executing `anonymous-statistics.sh`. Not a specialist, let's wait for @Ferroin reply.

File: daemon/analytics.c
Comment (thiagoftsm) [inline daemon/analytics.c]: Why `Next Generation` is in upper case while `legacy` is in lowercase?

File: daemon/analytics.c
Comment (thiagoftsm) [inline daemon/analytics.c:371]: Please, use `snprintfz` instead `sprintf`.

File: daemon/analytics.c
Comment (thiagoftsm) [inline daemon/analytics.c:389]: Please, move this semicolon for the previous line.

File: daemon/analytics.c
Comment (MrZammler) [reply daemon/analytics.c]: It was like that I think in web/api/info .... And so I left them the same, to be consistent with what we show there and in posthog. Now they are embedded probably in sql queries, so I'll leave them like that. :-)

File: daemon/analytics.c
Comment (thiagoftsm) [inline daemon/analytics.c:398]: The algorithm inside this `if` and the algorithm inside the previous `if` for `INSTALL_TYPE` has more or less the same code, the exception is one argument for the function `analytics_set_data_str`. I will give a suggestion to convert this for a function.

File: exporting/exporting_engine.c
Comment (thiagoftsm) [inline exporting/exporting_engine.c:18]: We are appending a `|` here, but we also append one inside the for loop when a condition is true, will we have problem when the parser receives something like `exporting||`?

File: daemon/analytics.c
Comment (MrZammler) [reply daemon/analytics.c]: Btw, this part has been moved to https://github.com/netdata/netdata/pull/11247 for a faster merge :-)

File: exporting/exporting_engine.c
Comment (MrZammler) [reply exporting/exporting_engine.c:18]: This is why @andrewm4894 is here :-) 

The string can be anything of `exporting|streaming|web`. We use the same notation in other parts, and posthog can extract each piece and create a new internal attribute from each (e.g. `has_https_streaming = true`).

In the end, I will leave the final say to Andy if he's ok with it, or if I should change the way we send those.

Comment (andrewm4894) [reply exporting/exporting_engine.c:18]: `exporting||` would be fine by me - the js plugin will split by "|" and just ignore any empty/null splits like "||" would give.

Comment (thiagoftsm) [reply exporting/exporting_engine.c:18]: Thank you @andrewm4894 to explain for us.

**Comment by andrewm4894** [comment]
@MrZammler can you share the personid (in posthog sense) for any dev/test machines you are using so i can eyeball how things look in posthog itself?

**Comment by MrZammler** [comment]
> @MrZammler can you share the personid (in posthog sense) for any dev/test machines you are using so i can eyeball how things look in posthog itself?

Hi Andy! You should see these attributes from `bbd2645a-a296-11eb-a8da-ac120315f513`.

File: daemon/analytics.c
Comment (MrZammler) [reply daemon/analytics.c:371]: Hmm, thanks Thiago. In that case though, it would be better to set `install_type_filename` to a fixed length (maybe `FILENAME_MAX`) and use that then in `snprintfz`. We are also pretty much sure that the `sprintf` won't exceed the length of `install_type_filename`, so it should be safe this way too? It's also the way I have used in most of the cases here...

If it's better to use a static length in such files related variables, I could change it to like that in all cases...

File: daemon/analytics.c
Comment (MrZammler) [reply daemon/analytics.c]: Thanks! changed with https://github.com/netdata/netdata/pull/11237/commits/8d59d9032b746dbaf1b643522303bfcf79b52a06

File: daemon/analytics.c
Comment (MrZammler) [reply daemon/analytics.c]: Or like this: https://github.com/netdata/netdata/pull/11237/commits/f03782563f5f426ef77da0102cb314a89c20516b if you're ok. :-)

Comment (thiagoftsm) [reply daemon/analytics.c]: I agree that for this case we cannot exceed the size, my suggestion to change the function was because the warning we have in the official documentation

```
Because sprintf() and vsprintf() assume an arbitrarily long string, callers must be careful not to overflow the actual space; this is often impossible to assure.  Note that the length of the strings produced  is  locale-de‐
       pendent and difficult to predict.  Use snprintf() and vsnprintf() instead (or asprintf(3) and vasprintf(3)).

       Code such as printf(foo); often indicates a bug, since foo may contain a % character.  If foo comes from untrusted user input, it may contain %n, causing the printf() call to write to memory and creating a security hole.
``` 

Thanks this warning, `snprintf` is always the safest option. 
I am OK with the suggested commit. :)

**Comment by andrewm4894** [comment]
i have added the code in the js plugin to handle the `config_https_available` string and break it out into flags. 

https://github.com/netdata/posthog-netdata-event-processing/pull/91/files

File: daemon/analytics.c
Comment (MrZammler) [reply daemon/analytics.c]: Done with https://github.com/netdata/netdata/pull/11237/commits/49e3c595e0d643e864df710bcc81f801f6768572 !

Comment (thiagoftsm) [reply daemon/analytics.c]: Thank you! :) I will test your PR now.

File: daemon/anonymous-statistics.sh.in
Comment (thiagoftsm) [reply daemon/anonymous-statistics.sh.in:26]: @Ferroin , please, can you give your opinion here?

File: daemon/analytics.c
Comment (thiagoftsm) [inline daemon/analytics.c:396]: We are passing as argument `4096`, but `buf` has only 256 bytes length. Whether you see the source code [here](https://github.com/netdata/netdata/blob/714d5a0fccf839ef4a7ff5401430fdca02f97300/libnetdata/libnetdata.c#L1081), you will verify possible problems, please, adjust the argument here or `buf` size.

File: daemon/analytics.c
Comment (thiagoftsm) [inline daemon/analytics.c:415]: Please, take a look to argument alignment.

File: daemon/analytics.c
Comment (thiagoftsm) [inline daemon/analytics.c:522]: The same here.

File: daemon/analytics.c
Comment (MrZammler) [reply daemon/analytics.c]: Embarrassing!...

File: daemon/analytics.c
Comment (MrZammler) [reply daemon/analytics.c]: This is how clang format formatted it... I've put it in one line, thanks.

**Comment by thiagoftsm** [comment]
@MrZammler thank you for your initial work :) , now that my initial concerns with code was removed, I compiled your PR and I ran it and I observed that there is one more thing necessary to adjust in your, code. I observed that my `error.log` is having this error:

```conf
2021-06-22 15:57:07: alarm-notify.sh: FATAL: All notification methods are disabled. Not sending notification for host 'hades', chart 'system.clock_sync_state' to 'silent' for 'system_clock_sync_state' = '0' for status 'WARNING'.
```

I do not have this kind of error when I am running the current master branch. After this error is reported, my `netdata` crashes and `valgrind` gives the following error:

```bash
==23032== Command: netdata -D
==23032== 
2021-06-22 15:56:43: netdata INFO  : MAIN : TIMEZONE: using the link of /etc/localtime: 'UTC'
2021-06-22 15:56:43: netdata INFO  : MAIN : TIMEZONE: fixed as 'UTC'
2021-06-22 15:56:43: netdata INFO  : MAIN : SIGNAL: Not enabling reaper
==23032== Thread 2:
==23032== Syscall param writev(vector[...]) points to uninitialised byte(s)
==23032==    at 0x59DE62D: writev (in /lib64/libc-2.33.so)
==23032==    by 0x4E0CDC9: ??? (in /usr/lib64/libuv.so.1.0.0)
==23032==    by 0x4E0DE9F: uv_write2 (in /usr/lib64/libuv.so.1.0.0)
==23032==    by 0x560992: spawn_process_cmd (spawn_client.c:160)
==23032==    by 0x560992: spawn_client (spawn_client.c:215)
==23032==    by 0x58C2E84: start_thread (in /lib64/libpthread-2.33.so)
==23032==    by 0x59E7B3E: clone (in /lib64/libc-2.33.so)
==23032==  Address 0x7e15a04 is 196 bytes inside a block of size 240 alloc'd
==23032==    at 0x483E77F: malloc (vg_replace_malloc.c:309)
==23032==    by 0x4309B0: mallocz (libnetdata.c:177)
==23032==    by 0x5608E5: spawn_process_cmd (spawn_client.c:142)
==23032==    by 0x5608E5: spawn_client (spawn_client.c:215)
==23032==    by 0x58C2E84: start_thread (in /lib64/libpthread-2.33.so)
==23032==    by 0x59E7B3E: clone (in /lib64/libc-2.33.so)
==23032== 
==23032== Thread 19 ANALYTICS:
==23032== Invalid read of size 8
==23032==    at 0x45BBAE: analytics_exporting_connectors_ssl (exporting_engine.c:10)
==23032==    by 0x41685D: analytics_https (analytics.c:414)
==23032==    by 0x416D79: analytics_gather_immutable_meta_data (analytics.c:542)
==23032==    by 0x4181D7: analytics_main (analytics.c:625)
==23032==    by 0x43FA59: thread_start (threads.c:184)
==23032==    by 0x58C2E84: start_thread (in /lib64/libpthread-2.33.so)
==23032==    by 0x59E7B3E: clone (in /lib64/libc-2.33.so)
==23032==  Address 0x30 is not stack'd, malloc'd or (recently) free'd
==23032== 
==23032== 
==23032== Process terminating with default action of signal 11 (SIGSEGV)
==23032==  Access not within mapped region at address 0x30
==23032==    at 0x45BBAE: analytics_exporting_connectors_ssl (exporting_engine.c:10)
==23032==    by 0x41685D: analytics_https (analytics.c:414)
==23032==    by 0x416D79: analytics_gather_immutable_meta_data (analytics.c:542)
==23032==    by 0x4181D7: analytics_main (analytics.c:625)
==23032==    by 0x43FA59: thread_start (threads.c:184)
==23032==    by 0x58C2E84: start_thread (in /lib64/libpthread-2.33.so)
==23032==    by 0x59E7B3E: clone (in /lib64/libc-2.33.so)
==23032==  If you believe this happened as a result of a stack
==23032==  overflow in your program's main thread (unlikely but
==23032==  possible), you can try to increase the size of the
==23032==  main thread stack using the --main-stacksize= flag.
==23032==  The main thread stack size used in this run was 8388608.
==23032== 
==23032== HEAP SUMMARY:
==23032==     in use at exit: 19,436,328 bytes in 326,335 blocks
==23032==   total heap usage: 559,524 allocs, 233,189 frees, 169,390,695 bytes allocated
==23032== 
==23032== LEAK SUMMARY:
==23032==    definitely lost: 6,565 bytes in 727 blocks
==23032==    indirectly lost: 0 bytes in 0 blocks
==23032==      possibly lost: 2,129,400 bytes in 4,440 blocks
==23032==    still reachable: 17,300,363 bytes in 321,168 blocks
==23032==                       of which reachable via heuristic:
==23032==                         length64           : 335,576 bytes in 598 blocks
==23032==         suppressed: 0 bytes in 0 blocks
==23032== Rerun with --leak-check=full to see details of leaked memory
==23032== 
==23032== Use --track-origins=yes to see where uninitialised values come from
==23032== For lists of detected and suppressed errors, rerun with: -s
==23032== ERROR SUMMARY: 2 errors from 2 contexts (suppressed: 0 from 0)
==23032== could not unlink /tmp/vgdb-pipe-from-vgdb-to-23032-by-root-on-hades
==23032== could not unlink /tmp/vgdb-pipe-to-vgdb-from-23032-by-root-on-hades
==23032== could not unlink /tmp/vgdb-pipe-shared-mem-vgdb-23032-by-root-on-hades
Segmentation fault
```

**Comment by MrZammler** [comment]
> @MrZammler thank you for your initial work :) , now that my initial concerns with code was removed, I compiled your PR and I ran it and I observed that there is one more thing necessary to adjust in your, code. I observed that my `error.log` is having this error:
> 
> ```
> 2021-06-22 15:57:07: alarm-notify.sh: FATAL: All notification methods are disabled. Not sending notification for host 'hades', chart 'system.clock_sync_state' to 'silent' for 'system_clock_sync_state' = '0' for status 'WARNING'.
> ```
> 
> I do not have this kind of error when I am running the current master branch. 

Hmmm, not sure about this.... But will check if something messes up there.... Thank you.

> After this error is reported, my `netdata` crashes and `valgrind` gives the following error:
> 
> Segmentation fault
> ```

Thanks Thiago, thanks... https://github.com/netdata/netdata/pull/11237/commits/2c2697e44364540377396799de7ff8505989b977 should fix that, but I'm not really sure about this code in general, i.e. if it will really pick up cases of exporters with ssl connection...

**Review by thiagoftsm — DISMISSED** [review]
After the last commit, the bug was fixed and I observed that the specific bug:

```
2021-06-22 15:57:07: alarm-notify.sh: FATAL: All notification methods are disabled. Not sending notification for host 'hades', chart 'system.clock_sync_state' to 'silent' for 'system_clock_sync_state' = '0' for status 'WARNING'.
```

does not have relationship with this PR. LGTM! Thank you @MrZammler !

File: daemon/anonymous-statistics.sh.in
Comment (MrZammler) [reply daemon/anonymous-statistics.sh.in:26]: Hi! We could just merge this, and check the results :-) If all we get is falses, we can re-check or remove it. We'll make sure it doesn't get into a dashboard, @andrewm4894 are you ok with this?

**Review by stelfrag — DISMISSED** [review]
Looks Ok

File: daemon/anonymous-statistics.sh.in
Comment (Ferroin) [reply daemon/anonymous-statistics.sh.in:26]: Sorry about the delayed response, this somehow slipped through the cracks.

In practice, this will only work if Netdata is running natively in the CI environment. It won’t work in containers regardless of the container management stack (part of the point of a container is to be decoupled from the host environment), and it may not work in other circumstances as well (for example, it may not work if netdata is started by something that scrubs the environment, like for example `sudo`).

Comment (ilyam8) [reply daemon/anonymous-statistics.sh.in:26]: > In practice, this will only work if Netdata is running natively in the CI environment.

So we need to remove the variable from the script/analytics, am I correct? It won't the way it is implemented.

Comment (stelfrag) [reply daemon/anonymous-statistics.sh.in:26]: If that is the case, let's take it out so we can merge this!

**Review by andrewm4894 — DISMISSED** [review]

File: daemon/anonymous-statistics.sh.in
Comment (MrZammler) [reply daemon/anonymous-statistics.sh.in]: OK, removed!

File: exporting/exporting_engine.c
Comment (ilyam8) [inline exporting/exporting_engine.c:14]: is that change related to removing `NETDATA_IN_CI`?

Comment (MrZammler) [reply exporting/exporting_engine.c:14]: No, just a quick fix :-) (The `|` will be added below).

Comment (ilyam8) [reply exporting/exporting_engine.c:14]: ok, asked because of the commit name 😄

**Review by thiagoftsm — APPROVED** [review]
I retest the PR after the last commit and everything is working. LGTM!

**Review by stelfrag — APPROVED** [review]

**Comment by MrZammler** [comment]
Thanks guys!

**Pull request merged**
